pkg/deploy/deploy.go

```diff
@@ -80,8 +88,101 @@ func (o *deployHandler) ApplyKubernetes(kubernetes v1alpha2.Component) error {
-	// * Make sure you inject the "component type" label once we implement exec.
-	return errors.New("exec command is not implemented for Deploy")
+	containerComps, err := generator.GetContainers(o.devfileObj, common.DevfileOptions{FilterByName: command.Exec.Component})
+	if err != nil {
+		return err
+	}
+	if len(containerComps) != 1 {
+		return fmt.Errorf("could not find the component")
+	}
+	containerComp := containerComps[0]
+
+	// Create a Kubernetes Job and use the container image referenced by command.Exec.Component
+	// Get the component for the command with command.Exec.Component
+	completionMode := batchv1.CompletionMode("Indexed")
+	job := batchv1.Job{
+		TypeMeta: generator.GetTypeMeta(kclient.JobsKind, kclient.JobsAPIVersion),
+		ObjectMeta: metav1.ObjectMeta{
+			// TODO: Check if the name is K8s valid.
+			Name: o.componentName + "-" + o.appName + "-" + command.Id, // TODO: Is there a function to return the standard odo names?
+		},
+		Spec: batchv1.JobSpec{
+			Template: corev1.PodTemplateSpec{
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Name:  containerComp.Name,
+							Image: containerComp.Image,
+							// TODO: Should we use the command and args already defined inside the container component?
+							// TODO: What if the 'sh' binary is not always available?
+							Command:         []string{"/bin/sh"},
+							Args:            getCmdline(command),
+							ImagePullPolicy: "IfNotPresent",
+						},
+					},
+					// Set the policy to `Never` so that it keeps the pod around and they can be used to debug.
+					RestartPolicy: "Never",
+				},
+			},
+			BackoffLimit:            pointer.Int32(2),
+			CompletionMode:          &completionMode,
+			TTLSecondsAfterFinished: pointer.Int32(60),
+		},
+		Status: batchv1.JobStatus{},
+	}
+
+	// Set labels and annotations
+	job.SetLabels(odolabels.GetLabels(o.componentName, o.appName, component.GetComponentRuntimeFromDevfileMetadata(o.devfileObj.Data.GetMetadata()), odolabels.ComponentDeployMode, false))
+	job.Annotations = map[string]string{}
+	odolabels.AddCommonAnnotations(job.Annotations)
+	odolabels.SetProjectType(job.Annotations, component.GetComponentTypeFromDevfileMetadata(o.devfileObj.Data.GetMetadata()))
+
+	log.Sectionf("Executing command in container (command: %s)", command.Id)
+	spinner := log.Spinnerf("Executing %q", command.Exec.CommandLine)
+	defer spinner.End(false)
+
+	var createdJob *batchv1.Job
+	createdJob, err = o.kubeClient.CreateJobs(job, "")
+	if err != nil {
+		return err
+	}
+
+	_, err = o.kubeClient.WaitForJobToComplete(createdJob)
+	spinner.End(err == nil)
+	if err != nil {
+		err = fmt.Errorf("failed to execute (command: %s)", command.Id)
+	}
+
+	jobLogs, logErr := o.kubeClient.GetJobLogs(createdJob, command.Exec.Component)
+	if logErr != nil {
+		log.Warningf("failed to fetch the logs of execution; cause: %s", logErr)
+	} else {
+		fmt.Println("Execution output:")
```

Comment:
Wouldn't it make more sense to display the logs only if the job failed? We could still `klog` the logs in all cases.

Reply:
https://github.com/redhat-developer/odo/pull/6579#issuecomment-1433103624